ldap/servers/slapd/haproxy.c

```diff
@@ -0,0 +1,457 @@
+ * return immediately with whatever data is available, preventing the thread from being
+ * stalled if the data is not immediately available.
+ */
+int set_nonblocking_if_needed(int fd, int *was_blocking) {
+    int flags = fcntl(fd, F_GETFL, 0);
+    if (flags == -1) {
+        slapi_log_err(SLAPI_LOG_ERR, "haproxy_receive", "Error getting socket flags: %s\n", strerror(errno));
+        return -1;
+    }
+
+    if (!(flags & O_NONBLOCK)) {
+        /* The socket is in blocking mode, set it to non-blocking mode */ 
+        if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) == -1) {
+            slapi_log_err(SLAPI_LOG_ERR, "haproxy_receive", "Error setting socket to non-blocking mode: %s\n", strerror(errno));
+            return -1;
+        }
+        *was_blocking = 1;
+    } else {
+        *was_blocking = 0;
+    }
+
+    return 0;
+}
+
+/** 
+ * Check if the socket was in blocking mode originally, and if it was, set it back to
+ * blocking mode. This is done to restore the original state of the socket, as other
+ * parts of the code might expect the socket to be in blocking mode and not be designed
+ * to handle the EAGAIN or EWOULDBLOCK errors that can occur in non-blocking mode.
+ */
+int set_blocking_if_needed(int fd, int was_blocking) {
+    int flags;
+    if (was_blocking) {
+        /* The socket was in blocking mode originally, set it back to blocking mode */
+        flags = fcntl(fd, F_GETFL, 0);
+        if (flags == -1) {
+            slapi_log_err(SLAPI_LOG_ERR, "haproxy_receive", "Error getting socket flags: %s\n", strerror(errno));
+            return -1;
+        }
+
+        if (fcntl(fd, F_SETFL, flags & ~O_NONBLOCK) == -1) {
+            slapi_log_err(SLAPI_LOG_ERR, "haproxy_receive", "Error setting socket back to blocking mode: %s\n", strerror(errno));
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+/**
+ * Function to receive and parse HAProxy headers, supporting both v1 and v2 of the protocol.
+ * 
+ * @param fd: The file descriptor of the socket from which to read.
+ * @param proxy_connection: A pointer to an integer to store the proxy connection status (0 or 1).
+ * @param pr_netaddr_from: A pointer to a PRNetAddr structure to store the source address info.
+ * @param pr_netaddr_dest: A pointer to a PRNetAddr structure to store the destination address info.
+ * 
+ * @return: Returns 0 on successful operation, -1 on error.
+ */
+int haproxy_receive(int fd, int *proxy_connection, PRNetAddr *pr_netaddr_from, PRNetAddr *pr_netaddr_dest)
+{
+    /* Buffer to store the header received from the HAProxy server */
+    char hdr[HAPROXY_HEADER_MAX_LEN + 1];
+    ssize_t recv_result = 0;
+    size_t hdr_len;
+    int was_blocking;
+
+    if (set_nonblocking_if_needed(fd, &was_blocking) == -1) {
+        return -1;
+    }
+    /* Attempt to receive the header from the HAProxy server */
+    recv_result = recv(fd, hdr, sizeof(hdr) - 1, MSG_PEEK);
+
+    if (set_blocking_if_needed(fd, was_blocking) == -1) {

```

Comment:
Thinking more about it: I think that resetting the mode at this point is ok, but we should not bother to switch again the mode around the second recv (We know that the data are present because we have already read them so the second recv will not block whatever the mode is)